# 게임 단말기 전환(피처폰에서 스마트폰으로) 사례분석
* 우리 회사는 '퍼즐컬렉션'외에도 '낚시랜드'라는 소셜게임을 오래전부터 운영하고 있습니다.<br> 이 게임을 운영하기 시작한 건 아직 스마트폰이 나오기 전이어서, 처음에는 피처폰에 특화된 시스템으로 운영되었습니다.<br> 그러다가 스마트폰이 나오고 빠른 속도로 보급되자 스마트폰을 위한 기능과, 기존 유저가 피처폰에서 스마트폰으로 갈아탈 수 있도록 아이디 이전 등의 기능을 추가로 제공하였습니다.<br> 이 기능이 추가되고부터 스마트폰 유저가 서서히 늘어나며 게임 전체의 유저수가 증가하기 시작했습니다.<br> 그런데 최근 '낚시랜드'의 유저수가 조금씩 감소하는 문제가 발생했습니다.<br> 원인을 여러모로 조사해본 결과, 피처폰의 유저수가 크게 감소하고 있다는 것이 밝혀졌습니다.

>* 단말기 시장의 변화<br>
최근 단말기 시장의 추세를 보면,피처폰 이용자는 줄어들고 스마트폰 이용자가 늘어나는 형태로 시장 환경이 변했습니다.<br> 이 흐름에서는 피처폰으로 게임을 이용하는 유저가 줄어드는건 피할 수 없는 상황입니다.<br> 다른 게임의 상황을 확인해본 바, 역시 별반 다르지 않다는 것을 알 수 있었습니다. <br> 그러나 다른게임에서는 피처폰 유저가 줄어든 대신 스마트폰 유저가 증가해서, 게임 전체 유저수는 거의 같은 수준을 유지하고 있었습니다.<br> 즉, '낚시랜드'는 다른 게임과 비교했을 때, 피처폰 유저수는 비슷하게 줄어든 반면, 스마트폰 유저수의 증가는 적었다고 할 수 있습니다.

## 현실의 모습 / 이상적인 모습
* 현실의 모습 : 피처폰 유저의 감소량에 비해 스마트폰 유저의 증가량이 적음
* 이상적인 모습 : 피처폰 유저 감소량 = 스마트폰 유저 증가량

## 문제 발견

* 문제
  - '낚시랜드'는 다른 게임과 비교했을 때, 피처폰 유저수는 비슷하게 줄어든 반면, 스마트폰 유저수의 증가는 적었다.
* 피처폰 유저의 탈퇴를 구조적으로 분석
  1. 피처폰에서 그냥 탈퇴
  2. 피처폰에서 스마트폰으로 갈아타기 위해 탈퇴 <br>
     -> ID를 이전해서 피처폰에서 쓰던 계정을 그대로 사용하는 유저<br>
     -> ID를 새로 만들어서 사용하는 유저<br>
     -> 자연히 탈퇴하는 유저<br>
<br>
* 문제 부가 설명<br>
  - 피처폰에서 스마트폰으로 ID 이전을 한 유저는 스마트폰 유저로 간주됨.
  - 스마트폰으로 이전할 때 새로 ID를 만들어서 게임을 이용하는 유저도 있을 것임.
    - 이 경우는 실제 유저가 같더라도 계정이 다르기 때문에 피처폰에서 갈아탄 기존 유저인지 알 방법이 없음.
    - 다만 이 경우는 정기적인 설문조사를 통해 이런 유저수는 극히 일부임이 밝혀짐.
  - 스마트폰으로 갈아탈 때 자연히 게임을 관두는 유저도 있을 것임.
    - 이 경우는 피처폰을 쓰다가 그냥 탈퇴한 것인지, 스마트폰으로 갈아타다보니 탈퇴한 것인지 알 방법이 없음.

  >위 내용을 가지고 '낚시랜드' 게임 기획자 개발자와 회의를 한 결과, '스마트폰으로 갈아타면서 게임을 떠난 유저 중 ID 이전에 실패해서 게임을 떠난 유저가 있지 않을까?' 라는 의견이 있었음.

  >ID 이전 기능은 스마트폰이 처음 나왔을 때 급히 만든 것이라 유저 인터페이스가 조금 어려웠을 것.<br> 그런데도 지금까지 별 문제가 없었던 것은 이 기능이 제공되었을 땐 스마트폰으로 갈아타던 유저들이 소위 '얼리어댑터'라서 유저 인터페이스가 조금 복잡하더라도 알아서 잘 이전했기 때문이라는 추측을 하게됨.

  >만약 피처폰에서 쓰던 계정을 스마트폰으로 이전하려고 했으나 잘 안되어서 게임을 떠난 것이라면, 조속히 시스템을 개선해야됨. 그러나 현재 ID 이전 시스템이 졸속으로 만들어진 관계로 상당한 예산이 필요할 것으로 예상됨. 또한 현재 진행중인 시스템 개선 사안도 있기 때문에 '일단 어떻게든 고치자' 는 식으로 문제를 해결할 수도 없는 상황.

  >일단 대충이라도 좋으니 가급적 빨리 'ID 이전 실패로 인한 탈퇴'가 전체 유저수 감소에 어느 정도 영향을 끼쳤는지 파악해서, 다음과 같은 식으로 액션을 취하기로 함.<br>
  -> 영향이 크다면 다른 사안을 중지해서라도 ID 이전 시스템을 개선한다.<br>
  -> 영향이 작다면 스마트폰 유저 획득을 위한 광고를 개선한다.

- 탈퇴 전의 이용 상황을 생각해보면 다음과 같은 특징이 있을 것.
  - 이용 횟수가 적음
    1. 스마트폰을 가지고 있지 않으며 자연히 게임을 떠남
    2. 스마트폰을 가지고 있으며 자연히 게임을 떠남
  - 이용 횟수가 많음
    3. 스마트폰에서는 새로 ID를 만들어서 게임을 이용함
    4. 스마트폰으로 ID를 이전해서 게임을 이용함
    5. 스마트폰으로 ID를 이전하려 했으나 잘 되지 않았음<br>
>3번은 설문조사를 통해 극히 소수임을 알고 있고, 실제 얻을 수 있는 데이터는 4번 밖에 없음.

정답을 포함하지 않은 데이터에 대해 억지로 모델을 만들어서 제안한다는 건 학술적으로 용납되지 않음.<br>
그러나 비즈니스에서는 매번 정답을 포함한 데이터를 손에 넣을 수 있다고 장담할 수 없음.
<br>
이번 사례와 같이 진정한 의미의 정답 데이터가 존재하지 않는 가운데, 어떻게든 비즈니스 상에서 의미가 있는 결론을 끌어내야 하는 상황은 드물지 않게 일어남.

* 가설에서처럼 ID 이전 실패로 인해 탈퇴한 유저수가 많을 경우<br>
  -> ID 이전 실패로 인해 탈퇴한 유저가 자연히 탈퇴한 유저에 비해 탈퇴한 달의 이용횟수가 훨씬 많을 것이라 예상됨<br>
  -> 탈퇴자 전체에 대한 'ID 이전 실패로 인한 탈퇴'의 비율이 높을 때는 탈퇴자 전체와 ID 이전에 성공한 유저 간에 1월의 이용 횟수가 크게 다르지 않을 것임.<br>
  -> 1월의 이용횟수가 크게 다르지 않다는 건 이용횟수를 가지고는 모델을 만들 수 없다는 얘기.<br>
  -> 반대로 '말하면 ID 이전 실패로 인한 탈퇴'가 적다면 이용횟수의 차이가 분명히 드러날 것이기에 모델을 만들 수 있음.<br>

>따라서 모델이 만들어진다면 ID 이전 실패로 인한 탈퇴자수가 그리 많지 않으므로 기존 ID 이전 기능의 복잡함이 크게 문제되지 않는다는 해석이 가능함.

### 데이터 분석 스토리 정리
* 문제
  - 피처폰 유저의 감소수가 스마트폰 유저의 증가수에 비해 많다.(사실)
* 사실
  - 피처폰 유저 중 스마트폰으로 갈아탈 때 ID를 새로 만들어서 사용하는 유저는 매우 적다. (설문조사로 알아낸 사실)
* 가설 
  - 피처폰에서 스마트폰으로 갈아탈 때 ID 이전에 실패해서 떠난 유저가 있다.(가설)
* 분석 기법과 결과를 바탕으로 한 액션
  - 자연 탈퇴한 유저는 점점 이용일이 줄어들다 떠나게 되는데, ID 이전에 실패해서 탈퇴한 유저는 쭉 이용하다가 갑자기 발길이 뚝 끊어졌을 것임.
  - 이를 바탕으로 탈퇴 직전 이용 상황 데이터를 활용해서 'ID 이전에 실패한 유저'와 '자연 탈퇴 유저'의 판별 모델을 구축
    - 유효한 모델이 만들어진 경우 : ID 이전 실패로 인해 탈퇴한 유저수가 적음<br>
   ->스마트폰 신규 유저를 대상으로 한 광고를 개선함
    - 유효한 모델이 만들어지지 않을 경우 : ID 이전 실패로 인해 탈퇴한 유저가 많을 가능성이 높음<br>
   -> ID 이전 기능을 개선해서 보다 간단하고 실수 없이 이전할 수 있도록 함

## 데이터 수집과 가공
### 분석에 필요한 데이터 수집

In [269]:
#필요한 라이브러리 불러오기
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings(action='ignore')
#CSV 파일 읽어 들이기
dau = pd.read_csv('section7-dau.csv')

In [270]:
dau.info()        
# region_month : 이용월
# region_day : 이용일
# app_name : 게임명
# user_id : 유저 ID
# device : 단말기 종류
print("============================================")
print(dau)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48988 entries, 0 to 48987
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   region_month  48988 non-null  object
 1   region_day    48988 non-null  object
 2   app_name      48988 non-null  object
 3   user_id       48988 non-null  int64 
 4   device        48988 non-null  object
dtypes: int64(1), object(4)
memory usage: 1.9+ MB
      region_month  region_day app_name   user_id device
0          2013-01  2013-01-01  game-02  10061580     FP
1          2013-01  2013-01-01  game-02  10154440     FP
2          2013-01  2013-01-01  game-02  10164762     SP
3          2013-01  2013-01-01  game-02  10165615     FP
4          2013-01  2013-01-01  game-02  10321356     FP
...            ...         ...      ...       ...    ...
48983      2013-02  2013-02-28  game-02   9839434     FP
48984      2013-02  2013-02-28  game-02   9874350     SP
48985      2013-02  2013-02-28  gam

In [271]:
#유저별로 ID 이전을 한 유저인지 아닌지 나타내는 데이터 정리
# FP = 피처폰(Feature Phone) 이용자 , SP = 스마트폰(Smart Phone)이용자
# + 31일치의 데이터 존재로 인한 중복값 제거
#MAU
mau = dau[['user_id','region_month','device']]
mau = mau.drop_duplicates(['region_month', 'device', 'user_id'])
mau=mau.sort_values('user_id')
mau=mau.reset_index(drop=True)

#1월 MAU
mau1=mau[mau['region_month']=='2013-01']
#2월 MAU
mau2=mau[mau['region_month']=='2013-02']

#1월 FP MAU
fp_mau1 = mau1[mau1.device == 'FP']
#2월 FP MAU
fp_mau2 = mau2[mau2.device == 'FP']

#1월 SP MAU
sp_mau1 = mau1[mau1.device == 'SP']
#2월 SP MAU
sp_mau2 = mau2[mau2.device == 'SP']

mau

user_id region_month device
0        397286      2013-01     FP
1        397286      2013-02     FP
2        413097      2013-02     FP
3        471341      2013-02     SP
4        471341      2013-01     FP
...         ...          ...    ...
2976  695000512      2013-02     SP
2977  695166076      2013-02     SP
2978  695257670      2013-02     SP
2979  695297242      2013-02     SP
2980  695385021      2013-02     SP

[2981 rows x 3 columns]

In [272]:
# 1월에 피처폰으로 이용하던 유저가 2월에 탈퇴했는지 혹은 계속 이용했는지 추출
mau2['is_access']=1
fp_mau1=pd.merge(fp_mau1,mau2[['user_id','is_access']],how = 'left', on = ['user_id'])
fp_mau1['is_access']=fp_mau1['is_access'].fillna(0).astype(int)
fp_mau1

user_id region_month device  is_access
0       397286      2013-01     FP          1
1       471341      2013-01     FP          1
2       503874      2013-01     FP          0
3       512250      2013-01     FP          1
4       513811      2013-01     FP          1
..         ...          ...    ...        ...
901   61540141      2013-01     FP          0
902   61570858      2013-01     FP          0
903   61613038      2013-01     FP          0
904   61975338      2013-01     FP          1
905  677996362      2013-01     FP          0

[906 rows x 4 columns]

In [273]:
#1월에 피처폰으로 이용하던 유저가 2월에 이용한 단말기가 피처폰인지 스마트폰인지 구별하기
fp_mau2['is_fp']=1
sp_mau2['is_sp']=1
fp_mau1 = pd.merge(fp_mau1, fp_mau2[['user_id', 'is_fp']],on='user_id', how = 'left')
fp_mau1['is_fp'] = fp_mau1['is_fp'].fillna(0).astype(int)
fp_mau1 = pd.merge(fp_mau1, sp_mau2[['user_id', 'is_sp']],on='user_id', how = 'left')
fp_mau1['is_sp'] = fp_mau1['is_sp'].fillna(0).astype(int)
fp_mau1

user_id region_month device  is_access  is_fp  is_sp
0       397286      2013-01     FP          1      1      0
1       471341      2013-01     FP          1      0      1
2       503874      2013-01     FP          0      0      0
3       512250      2013-01     FP          1      1      0
4       513811      2013-01     FP          1      1      0
..         ...          ...    ...        ...    ...    ...
901   61540141      2013-01     FP          0      0      0
902   61570858      2013-01     FP          0      0      0
903   61613038      2013-01     FP          0      0      0
904   61975338      2013-01     FP          1      1      0
905  677996362      2013-01     FP          0      0      0

[906 rows x 6 columns]

In [274]:
#로지스틱 회귀분석에 필요한 데이터 추출
#1월에는 피처폰으로 이용했는데 2월에는 이용하지 않았거나 혹은 2월에 스마트폰으로 이용한 유저
fp_mau1 = fp_mau1[(fp_mau1.is_access == 0 ) | (fp_mau1.is_sp == 1)]
fp_mau1

user_id region_month device  is_access  is_fp  is_sp
1       471341      2013-01     FP          1      0      1
2       503874      2013-01     FP          0      0      0
10     1073544      2013-01     FP          0      0      0
11     1073864      2013-01     FP          0      0      0
13     1163733      2013-01     FP          1      0      1
..         ...          ...    ...        ...    ...    ...
900   61132242      2013-01     FP          0      0      0
901   61540141      2013-01     FP          0      0      0
902   61570858      2013-01     FP          0      0      0
903   61613038      2013-01     FP          0      0      0
905  677996362      2013-01     FP          0      0      0

[252 rows x 6 columns]

>분석하면서 깨달은 것<br>
fp_mau1 = fp_mau1[fp_mau1.is_fp==0] 이렇게 하면 안되나? -> 안됨<br>
1월에 피처폰을 이용하고 2월에도 피처폰으로 이용하는 고객은 1월에 꾸준히 서비스를<br> 유지한 유저로 2월에 스마트폰으로 ID 이전을 한 고객과 같은 집단임<br>

In [275]:
#날짜별 게임 이용 상황 데이터 정리하기
fp_dau1 = dau[(dau.device == "FP") & (dau.region_month == "2013-01")]
fp_dau1['is_access'] = 1
fp_dau1 = fp_dau1.pivot_table(index=['user_id'],columns = 'region_day',values = 'is_access').reset_index()
fp_dau1 = fp_dau1.fillna(0).astype(int)
fp_dau1

region_day    user_id  2013-01-01  2013-01-02  2013-01-03  2013-01-04  \
0              397286           1           1           1           1   
1              471341           1           1           1           1   
2              503874           1           0           0           0   
3              512250           1           1           1           1   
4              513811           0           0           0           0   
..                ...         ...         ...         ...         ...   
881          61540141           0           0           0           0   
882          61570858           0           0           0           0   
883          61613038           0           0           0           0   
884          61975338           0           0           0           0   
885         677996362           1           0           0           0   

region_day  2013-01-05  2013-01-06  2013-01-07  2013-01-08  2013-01-09  ...  \
0                    1           1           1           1           1  ...   
1                    0           0           0           0           0  ...   
2                    0           0           0           0           0  ...   
3                    1           1           1           1           1  ...   
4                    0           0           0           0           0  ...   
..                 ...         ...         ...         ...         ...  ...   
881                  0           0           0           0           0  ...   
882                  0           0           0           0           0  ...   
883                  0           0           0           0           0  ...   
884                  0           0           0           0           0  ...   
885                  0           0           0           0           0  ...   

region_day  2013-01-22  2013-01-23  2013-01-24  2013-01-25  2013-01-26  \
0                    1           1           1           1           1   
1                    0           0           0           0           0   
2                    0           0           0           0           0   
3                    1           1           1           1           1   
4                    1           0           0           0           0   
..                 ...         ...         ...         ...         ...   
881                  0           0           1           1           1   
882                  0           0           0           0           0   
883                  0           0           0           0           0   
884                  0           0           0           0           0   
885                  0           0           0           0           0   

region_day  2013-01-27  2013-01-28  2013-01-29  2013-01-30  2013-01-31  
0                    1           1           1           1           1  
1                    0           0           0           0           0  
2                    0           0           0           0           0  
3                    1           1           1           1           1  
4                    0           1           1           0           1  
..                 ...         ...         ...         ...         ...  
881                  1           0           0           0           0  
882                  0           0           0           0           0  
883                  0           0           0           0           0  
884                  0           0           0           1           1  
885                  0           0           0           0           0  

[886 rows x 32 columns]

In [276]:
#2월에 스마트폰으로 이용한 유저 데이터 결합하기
#1월에 피처폰으로 이용했던 유저들에 대해, 탈퇴한 유저들은 유저군 '0'으로, 
#스마트폰으로 옮겨서 이용한 유저들은 유저군 '1'로 구분되어 있음
fp_dau_cast = pd.merge(fp_dau1, fp_mau1[['user_id', 'is_sp']], on = ['user_id'], how = 'inner')
fp_dau_cast = fp_dau_cast.fillna(0).astype(int)
fp_dau_cast

user_id  2013-01-01  2013-01-02  2013-01-03  2013-01-04  2013-01-05  \
0       471341           1           1           1           1           0   
1       503874           1           0           0           0           0   
2      1073544           0           0           0           0           0   
3      1073864           0           0           0           0           0   
4      1163733           1           1           0           0           0   
..         ...         ...         ...         ...         ...         ...   
247   61132242           0           0           0           0           0   
248   61540141           0           0           0           0           0   
249   61570858           0           0           0           0           0   
250   61613038           0           0           0           0           0   
251  677996362           1           0           0           0           0   

     2013-01-06  2013-01-07  2013-01-08  2013-01-09  ...  2013-01-23  \
0             0           0           0           0  ...           0   
1             0           0           0           0  ...           0   
2             0           0           0           0  ...           1   
3             0           0           0           0  ...           0   
4             0           0           0           0  ...           1   
..          ...         ...         ...         ...  ...         ...   
247           0           0           1           1  ...           0   
248           0           0           0           0  ...           0   
249           0           0           0           0  ...           0   
250           0           0           0           0  ...           0   
251           0           0           0           0  ...           0   

     2013-01-24  2013-01-25  2013-01-26  2013-01-27  2013-01-28  2013-01-29  \
0             0           0           0           0           0           0   
1             0           0           0           0           0           0   
2             1           1           0           0           0           0   
3             0           0           0           0           0           0   
4             1           1           1           1           1           0   
..          ...         ...         ...         ...         ...         ...   
247           0           0           0           0           0           0   
248           1           1           1           1           0           0   
249           0           0           0           0           0           0   
250           0           0           0           0           0           0   
251           0           0           0           0           0           0   

     2013-01-30  2013-01-31  is_sp  
0             0           0      1  
1             0           0      0  
2             0           0      0  
3             0           0      0  
4             0           0      1  
..          ...         ...    ...  
247           0           0      0  
248           0           0      0  
249           0           0      0  
250           0           0      0  
251           0           0      0  

[252 rows x 33 columns]

In [277]:
fp_dau_cast['is_sp'].value_counts()
# 탈퇴 : 190명 / ID 이전 : 62명

0    190
1     62
Name: is_sp, dtype: int64

### 모델 구축과 모델 검증 _ 로지스틱회귀분석

In [278]:
# 로지스틱 회귀 모형 생성
import statsmodels.api as sm
X = fp_dau_cast.drop(['user_id','is_sp'], axis=1)
y = fp_dau_cast['is_sp']

# 상수항 추가
X = sm.add_constant(X)

# 모델 생성
logit_model = sm.Logit(y, X)
result = logit_model.fit()

# 모델 결과 확인
result.summary()

Optimization terminated successfully.
         Current function value: 0.222882
         Iterations 9


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  is_sp   No. Observations:                  252
Model:                          Logit   Df Residuals:                      220
Method:                           MLE   Df Model:                           31
Date:                Sat, 01 Apr 2023   Pseudo R-squ.:                  0.6005
Time:                        04:46:34   Log-Likelihood:                -56.166
converged:                       True   LL-Null:                       -140.60
Covariance Type:            nonrobust   LLR p-value:                 6.629e-21
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -4.2551      0.592     -7.192      0.000      -5.415      -3.096
2013-01-01     1.9873      0.804      2.471      0.013       0.411       3.564
2013-01-02     0.3245      1.070      0.303      0.762      -1.774       2.423
2013-01-03     0.3871      0.940      0.412      0.681      -1.455       2.229
2013-01-04     2.0520      0.836      2.455      0.014       0.414       3.690
2013-01-05    -1.7603      1.202     -1.464      0.143      -4.116       0.596
2013-01-06    -0.6650      1.172     -0.567      0.570      -2.962       1.632
2013-01-07     2.0119      1.120      1.796      0.072      -0.184       4.207
2013-01-08     1.2009      1.352      0.888      0.375      -1.450       3.852
2013-01-09    -0.4430      1.187     -0.373      0.709      -2.769       1.883
2013-01-10    -3.2135      1.555     -2.067      0.039      -6.261      -0.166
2013-01-11     1.4348      1.224      1.172      0.241      -0.964       3.834
2013-01-12    -0.6478      1.530     -0.423      0.672      -3.647       2.351
2013-01-13     1.7741      1.144      1.551      0.121      -0.467       4.015
2013-01-14    -1.1306      1.248     -0.906      0.365      -3.576       1.315
2013-01-15    -1.8019      1.303     -1.382      0.167      -4.357       0.753
2013-01-16     0.4899      1.167      0.420      0.675      -1.798       2.778
2013-01-17    -0.4389      1.222     -0.359      0.719      -2.834       1.956
2013-01-18     1.4439      1.145      1.261      0.207      -0.801       3.688
2013-01-19    -0.5965      1.196     -0.499      0.618      -2.941       1.748
2013-01-20    -0.3477      1.400     -0.248      0.804      -3.092       2.397
2013-01-21     0.9714      1.254      0.775      0.439      -1.487       3.430
2013-01-22     1.1868      1.123      1.057      0.291      -1.014       3.388
2013-01-23    -1.5476      1.233     -1.256      0.209      -3.964       0.868
2013-01-24     1.8509      1.102      1.680      0.093      -0.309       4.011
2013-01-25     0.1355      1.532      0.088      0.930      -2.868       3.139
2013-01-26     0.3079      1.429      0.215      0.829      -2.494       3.109
2013-01-27     0.3539      1.295      0.273      0.785      -2.184       2.892
2013-01-28    -0.4434      1.854     -0.239      0.811      -4.076       3.190
2013-01-29     1.6132      1.055      1.529      0.126      -0.455       3.682
2013-01-30     2.0081      1.211      1.658      0.097      -0.366       4.382
2013-01-31     1.5179      1.182      1.284      0.199      -0.799       3.835
==============================================================================
"""

In [279]:
# chatGPT로 구현한 로지스틱 회귀 단계적 선택법 (AIC를 기준으로)
def stepwise_selection_aic(X, y, initial_list=[], verbose=True, direction='both'):
    included = list(initial_list)
    best_aic = float('inf')  # 초기 값 설정
    while True:
        changed=False
        excluded = list(set(X.columns)-set(included))
        new_aic = pd.Series(index=excluded)
        for new_column in excluded:
            if direction == 'forward':
                model = sm.Logit(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit(disp=0)
            else:
                model = sm.Logit(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit(disp=0)
                if best_aic is None or model.aic < best_aic:
                    direction = 'forward'
                else:
                    direction = 'backward'
                    included.remove(new_column)
            new_aic[new_column] = model.aic
        min_aic = new_aic.min()
        if min_aic < best_aic:
            best_aic = min_aic
            best_feature = new_aic.idxmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print(f'Add {best_feature} with AIC {best_aic:.2f}')
        if not changed:
            break
    while True:
        changed=False
        model = sm.Logit(y, sm.add_constant(pd.DataFrame(X[included]))).fit(disp=0)
        pvalues = model.pvalues.iloc[1:]
        worst_aic = model.aic
        worst_feature = None  # 초기화
        for col in pvalues.index:
            temp = included.copy()
            temp.remove(col)
            model = sm.Logit(y, sm.add_constant(pd.DataFrame(X[temp]))).fit(disp=0)
            if model.aic < worst_aic:
                worst_aic = model.aic
                worst_feature = col
        if worst_feature is not None and worst_aic < model.aic:
            included.remove(worst_feature)
            changed=True
            if verbose:
                print(f'Drop {worst_feature} with AIC {worst_aic:.2f}')
        if not changed:
            break
    return included


In [280]:
y_list = stepwise_selection_aic(X, y, direction='both')
print(y_list)

Add 2013-01-28 with AIC 183.78
Add 2013-01-04 with AIC 166.28
Add 2013-01-22 with AIC 157.72
Add 2013-01-30 with AIC 155.04
Add 2013-01-27 with AIC 154.20
Add 2013-01-01 with AIC 153.20
Add 2013-01-29 with AIC 151.91
Add 2013-01-10 with AIC 150.41
Add 2013-01-08 with AIC 148.74
Drop 2013-01-28 with AIC 146.74
Drop 2013-01-27 with AIC 145.47
['2013-01-04', '2013-01-22', '2013-01-30', '2013-01-01', '2013-01-29', '2013-01-10', '2013-01-08']


In [281]:
y_list.sort()
# 로지스틱 회귀 재실행
X = fp_dau_cast[y_list]
y = fp_dau_cast['is_sp']

# 상수항 추가
X = sm.add_constant(X)
# 모델 생성
logit_model = sm.Logit(y, X)
result = logit_model.fit()

# 모델 결과 확인
print('AIC : ',result.aic)
result.summary()

Optimization terminated successfully.
         Current function value: 0.256888
         Iterations 7
AIC :  145.47130925606186


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  is_sp   No. Observations:                  252
Model:                          Logit   Df Residuals:                      244
Method:                           MLE   Df Model:                            7
Date:                Sat, 01 Apr 2023   Pseudo R-squ.:                  0.5396
Time:                        04:46:35   Log-Likelihood:                -64.736
converged:                       True   LL-Null:                       -140.60
Covariance Type:            nonrobust   LLR p-value:                 1.761e-29
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.6041      0.437     -8.246      0.000      -4.461      -2.747
2013-01-01     1.6270      0.567      2.871      0.004       0.516       2.738
2013-01-04     1.7917      0.587      3.051      0.002       0.641       2.943
2013-01-08     1.9219      0.790      2.434      0.015       0.374       3.470
2013-01-10    -2.7657      1.017     -2.719      0.007      -4.759      -0.772
2013-01-22     1.7149      0.590      2.905      0.004       0.558       2.872
2013-01-29     1.6546      0.659      2.512      0.012       0.364       2.945
2013-01-30     1.5682      0.732      2.143      0.032       0.134       3.003
==============================================================================
"""

In [282]:
# 잔차 추출
residuals = result.resid_dev
mean = round(np.mean(residuals), 4)
stddev = round(np.std(residuals), 4)
minimum = round(np.min(residuals), 4)
first_quantile = round(np.percentile(residuals, 25), 4)
median = round(np.median(residuals), 4)
third_quantile = round(np.percentile(residuals, 75), 4)
maximum = round(np.max(residuals), 4)
df = pd.DataFrame({'Mean': [mean],
                   'Std. Dev.': [stddev],
                   'Minimum': [minimum],
                   '25th percentile': [first_quantile],
                   'Median': [median],
                   '75th percentile': [third_quantile],
                   'Maximum': [maximum]})
df

Mean  Std. Dev.  Minimum  25th percentile  Median  75th percentile  \
0 -0.1006     0.7097  -2.2056          -0.4954 -0.2317          -0.0585   

   Maximum  
0   2.6948

In [283]:
# Coefficients
tables = result.summary().tables[1]
tables

<class 'statsmodels.iolib.table.SimpleTable'>

In [284]:
# 모델 검증
y_pred = result.predict(X)
y_prob = result.predict(X)
y_pred = [1 if i > 0.5 else 0 for i in y_pred] # 확률 0.5를 기준으로 1 또는 0으로 변환

fp_dau_cast['y_prob']=y_prob
fp_dau_cast['y_pred']=y_pred
fp_dau_cast

user_id  2013-01-01  2013-01-02  2013-01-03  2013-01-04  2013-01-05  \
0       471341           1           1           1           1           0   
1       503874           1           0           0           0           0   
2      1073544           0           0           0           0           0   
3      1073864           0           0           0           0           0   
4      1163733           1           1           0           0           0   
..         ...         ...         ...         ...         ...         ...   
247   61132242           0           0           0           0           0   
248   61540141           0           0           0           0           0   
249   61570858           0           0           0           0           0   
250   61613038           0           0           0           0           0   
251  677996362           1           0           0           0           0   

     2013-01-06  2013-01-07  2013-01-08  2013-01-09  ...  2013-01-25  \
0             0           0           0           0  ...           0   
1             0           0           0           0  ...           0   
2             0           0           0           0  ...           1   
3             0           0           0           0  ...           0   
4             0           0           0           0  ...           1   
..          ...         ...         ...         ...  ...         ...   
247           0           0           1           1  ...           0   
248           0           0           0           0  ...           1   
249           0           0           0           0  ...           0   
250           0           0           0           0  ...           0   
251           0           0           0           0  ...           0   

     2013-01-26  2013-01-27  2013-01-28  2013-01-29  2013-01-30  2013-01-31  \
0             0           0           0           0           0           0   
1             0           0           0           0           0           0   
2             0           0           0           0           0           0   
3             0           0           0           0           0           0   
4             1           1           1           0           0           0   
..          ...         ...         ...         ...         ...         ...   
247           0           0           0           0           0           0   
248           1           1           0           0           0           0   
249           0           0           0           0           0           0   
250           0           0           0           0           0           0   
251           0           0           0           0           0           0   

     is_sp    y_prob  y_pred  
0        1  0.453772       0  
1        0  0.121627       0  
2        0  0.001710       0  
3        0  0.026491       0  
4        1  0.434814       0  
..     ...       ...     ...  
247      0  0.156807       0  
248      0  0.026491       0  
249      0  0.026491       0  
250      0  0.026491       0  
251      0  0.121627       0  

[252 rows x 35 columns]

In [285]:
from sklearn.metrics import precision_score, recall_score, f1_score, confusion_matrix
f1 = f1_score(fp_dau_cast['is_sp'], fp_dau_cast['y_pred'])

# confusion matrix 계산
tn, fp, fn, tp = confusion_matrix(fp_dau_cast['is_sp'], fp_dau_cast['y_pred']).ravel()

print("Confusion Matrix:")
print(confusion_matrix(fp_dau_cast['is_sp'], fp_dau_cast['y_pred']))

print("F1 Score: {:.4f}".format(f1))

Confusion Matrix:
[[185   5]
 [ 22  40]]
F1 Score: 0.7477


## 데이터 분석 결론
                탈퇴(예측)  ID 이전(예측)
탈퇴(실제)       　　　    185       　　　　     5<br>
ID 이전 (실제)      　　  22        　　　　    40<br>
F1 Score : 0.7477 -> 신뢰할 만한 모델이라고 할 수 있음.

>이번 모델에서 1(ID 이전)로 예측되었으며 실제로 탈퇴한 5명이 ID 이전에 실패한 유저라고 생각할 수 있다.<br>
과거의 이용상황을 봤을 때 ID 이전을 해서 계속 이용해야 하는데, 실제로는 탈퇴해버린 유저들이야말로 ID 이전에 실패해서 게임을 떠나버린 유저군일 것이다.

>처음의 가설이었던 'ID 이전에 실패해서 탈퇴'한 유저의 영향은 작은 것으로 밝혀짐

## 액션
> 전체 탈퇴 유저수에서 ID를 이전하는데 실패해서 게임을 떠난 유저 비율이 높지 않음이 밝혀졌음.<br>
따라서 액션으로는 처음에 정한 대로 스마트폰 신규 유저를 향한 광고를 재검토하기로 결정

## 정리
#### 로지스틱 회귀분석을 이용한 분석
>실제 비지니스에서는 데이터 수집 단계에서 진정한 의미의 정답 데이터가 존재하지 않거나, 실시하고자 하는 분석 기법에 걸맞은 데이터가 존재하지 않는등 번거롭지만 직접 데이터를 가공해야 하는 경우가 많이 생긴다.<br> 그럴 경우 데이터 분석이나 모델 구축보다 그전에 데이터를 수집하고 가공하는 데 많은 노력이 필요하다는 것을 몸소 느꼈고 실제 비즈니스에서 데이터를 분석할 때의 어려움을 느꼈다.